src/crimson/osd/ops_executer.cc

```diff
@@ -723,6 +724,114 @@ version_t OpsExecuter::get_last_user_version() const
   return pg->get_last_user_version();
 }
 
+void OpsExecuter::make_writeable(std::vector<pg_log_entry_t>& log_entries)
+{
+  const hobject_t& soid = obc->obs.oi.soid;
+  logger().debug("{} {} snapset={} snapc={}",
+                 __func__, soid,
+                 obc->ssc->snapset, snapc);
+
+  // clone?
+  if (head_existed &&                            // old obs.exists
+      snapc.snaps.size() &&                      // there are snaps
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing obj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at it's creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+  // existing snaps are stored desendingly in snapc,
+  // cloned_snaps vector will hold all the snaps stored until snapset.seq
+   std::vector<snapid_t> cloned_snaps = [&] {
+      auto last = std::find_if(
+        std::begin(snapc.snaps), std::end(snapc.snaps),
+        [&](snapid_t snap_id) { return snap_id <= obc->ssc->snapset.seq; });
+      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
+    }();
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    auto snap_oi = prepare_clone(coid);
+
+    // make clone
+    do_write_op([this] (auto& backend, auto& os, auto& txn) {
+      return backend.clone(os, clone_obc->obs, txn);
+    }, true);
+    // TODO: maybe cache:
+    // encode(snap_oi, bv, get_osdmap()->get_features(CEPH_ENTITY_TYPE_OSD, nullptr));
+    // txn.setattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, OI_ATTR, bv);
+    // txn.rmattr(coll->get_cid(), ghobject_t{obc->obs.oi.soid}, SS_ATTR);
```

Comment:
This probably needs to be uncommented?

Reply:
Updated.

Reply:
> Getting really close! What testing have you done so far?

Mostly local testing using the rados tool. 
> If you've got some teuthology tests you used, I think the next step would be to split those out into a crimson smoke suite and start gating all crimson PRs on it.

I agree, I will start working on a minimal suite to run against.